In [2]:
# Imports and pip installations (if needed)


# Part 1: Load the dataset

In [3]:
# Load the given datasets
import pandas as pd
import math
from ucimlrepo import fetch_ucirepo 
pd.set_option('display.max_columns', None)


# fetch dataset 
chronic_kidney_disease = fetch_ucirepo(id=336) 
  
# data (as pandas dataframes) 
X = chronic_kidney_disease.data.features 
y = chronic_kidney_disease.data.targets 
  
# metadata 
print(chronic_kidney_disease.metadata) 
  
# variable information 
print(chronic_kidney_disease.variables) 

# print(chronic_kidney_disease.data)

# chronic_kidney_disease is a pandas dataframe
# display(chronic_kidney_disease)
#print(chronic_kidney_disease)


df_numerical = pd.read_csv('chronic_kidney_disease_numerical.csv')
df_categorical = pd.read_csv('chronic_kidney_disease_categorical.csv')


{'uci_id': 336, 'name': 'Chronic Kidney Disease', 'repository_url': 'https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease', 'data_url': 'https://archive.ics.uci.edu/static/public/336/data.csv', 'abstract': 'This dataset can be used to predict the chronic kidney disease and it can be collected from the hospital nearly 2 months of period.', 'area': 'Other', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 400, 'num_features': 24, 'feature_types': ['Real'], 'demographics': ['Age'], 'target_col': ['class'], 'index_col': None, 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 2015, 'last_updated': 'Mon Mar 04 2024', 'dataset_doi': '10.24432/C5G020', 'creators': ['L. Rubini', 'P. Soundarapandian', 'P. Eswaran'], 'intro_paper': None, 'additional_info': {'summary': 'We use the following representation to collect the dataset\r\n                        age\t\t-\tage\t\r\n\t\t\tbp\t\t-\tblood pressure\r\n\t\t\tsg\t

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

Each row of the data has a value for each of the 24 different variables (or columns). Each variable represents one factor that can be connected to kidney disease. Some of these columns are categorical - like specific gravity (sg), albumin (al), and sugar (su). Some of the columns are numerical, like age, blood pressure, and white/red blood cell count. The dataset further divides the numerical columns into "Integer" and "Continuous" (probably floating-point) values. Some values are also binary.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_duplicates = len(df_numerical) - len(df_numerical.drop_duplicates())
print("Number of duplicate rows (numerical dataset): " + str(num_duplicates))

# Drop these duplicate rows
if (num_duplicates != 0):
    df_numerical = df_numerical.drop_duplicates(keep=False)
    print("Duplicate rows dropped")


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
num_duplicates = len(df_categorical) - len(df_categorical.drop_duplicates())
print("Number of duplicate rows (categorical dataset): " + str(num_duplicates))

if (num_duplicates != 0):
    df_categorical = df_categorical.drop_duplicates(keep=False)
    print("Duplicate rows dropped")


Number of duplicate rows (numerical dataset): 30
Duplicate rows dropped
Number of duplicate rows (categorical dataset): 30
Duplicate rows dropped


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [5]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df_merged = pd.merge(df_categorical, df_numerical, on=["unique_id"])
print(len(df_merged))
#Print the combined dataset

df_merged


361


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd,35.0,70.0,82.0,36.0,1.1,150.0,3.5,14.5,52.0,9400.0,6.1
1,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd,65.0,70.0,139.0,29.0,1.0,NaN,NaN,10.5,32.0,NaN,NaN
2,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd,5.0,NaN,NaN,16.0,0.7,138.0,3.2,8.1,NaN,NaN,NaN
3,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,42.0,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
4,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
356,397388,2.0,2.0,normal,normal,notpresent,present,yes,no,yes,good,no,no,ckd,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2
357,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,69.0,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
358,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,43.0,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9
359,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd,40.0,80.0,76.0,162.0,9.6,141.0,4.9,7.6,24.0,3800.0,2.8


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value
num_incomplete_rows = df_merged.isnull().any(axis=1).sum()
percent = (num_incomplete_rows / len(df_merged)) * 100

# Print %
print(str(num_incomplete_rows) + " incomplete rows / " + str(len(df_merged)) + " total rows = " + str(percent) + "%")

# Drop these rows from the dataset
df_merged.dropna(inplace=True)

# Print the Dataset
df_merged


223 incomplete rows / 361 total rows = 61.772853185595565%


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
3,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,42.0,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
4,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2
6,901528,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,73.0,80.0,118.0,44.0,0.7,137.0,3.5,14.8,45.0,9300.0,4.7
7,433613,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,50.0,80.0,97.0,40.0,0.6,150.0,4.5,14.2,48.0,10500.0,5.0
8,621332,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,poor,no,no,ckd,59.0,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
355,996076,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2
356,397388,2.0,2.0,normal,normal,notpresent,present,yes,no,yes,good,no,no,ckd,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2
357,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,69.0,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
358,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,43.0,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9


## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df_merged.sort_values(by="Target", inplace=True)
df_merged = df_merged.reset_index()

# Print the dataset
df_merged

,index,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,285,546225,1.0,0.0,normal,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,14600.0,3.2
1,180,828592,3.0,0.0,normal,abnormal,present,present,yes,yes,yes,good,no,no,ckd,71.0,70.0,219.0,82.0,3.6,133.0,4.4,10.4,33.0,5600.0,3.6
2,54,746487,4.0,2.0,normal,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd,59.0,100.0,252.0,40.0,3.2,137.0,4.7,11.2,30.0,26400.0,3.9
3,245,197452,4.0,1.0,abnormal,normal,notpresent,notpresent,no,no,no,good,no,no,ckd,56.0,70.0,210.0,26.0,1.7,136.0,3.8,16.1,52.0,12500.0,5.6
4,268,923613,4.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,no,poor,yes,no,ckd,64.0,60.0,239.0,58.0,4.3,137.0,5.4,9.5,29.0,7500.0,3.4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
133,115,489160,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0
134,112,530138,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0
135,102,235579,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,46.0,60.0,102.0,27.0,0.7,142.0,4.9,13.2,44.0,11000.0,5.4
136,149,155117,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,68.0,60.0,125.0,41.0,1.1,139.0,3.8,17.4,50.0,6700.0,6.1


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dummies = pd.get_dummies(df_merged[categorical_columns]).astype(int)
df_merged = pd.concat([df_merged.drop(columns=categorical_columns), dummies], axis=1)



# Print the dataset
df_merged

,index,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,285,546225,1.0,0.0,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,14600.0,3.2,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
1,180,828592,3.0,0.0,71.0,70.0,219.0,82.0,3.6,133.0,4.4,10.4,33.0,5600.0,3.6,0,1,1,0,0,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,1,0
2,54,746487,4.0,2.0,59.0,100.0,252.0,40.0,3.2,137.0,4.7,11.2,30.0,26400.0,3.9,0,1,0,1,1,0,1,0,0,1,0,1,1,0,0,1,0,1,1,0,1,0
3,245,197452,4.0,1.0,56.0,70.0,210.0,26.0,1.7,136.0,3.8,16.1,52.0,12500.0,5.6,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0
4,268,923613,4.0,1.0,64.0,60.0,239.0,58.0,4.3,137.0,5.4,9.5,29.0,7500.0,3.4,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
133,115,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
134,112,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
135,102,235579,0.0,0.0,46.0,60.0,102.0,27.0,0.7,142.0,4.9,13.2,44.0,11000.0,5.4,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
136,149,155117,0.0,0.0,68.0,60.0,125.0,41.0,1.1,139.0,3.8,17.4,50.0,6700.0,6.1,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
N = df_merged.index.size
for column in numerical_columns:
    mean = df_merged[column].mean()
    print("mean of " + column + " is " + str(mean))
    sum = 0
    for index, row in df_merged.iterrows():
        x_i = row[column]
        sum += (x_i - mean) * (x_i - mean)
    standard_deviation = math.sqrt(sum / N)
    print("standard deviation of " + column + " is " + str(standard_deviation))
    range_min = mean - (3 * standard_deviation)
    range_max = mean + (3 * standard_deviation)
    print("acceptable range for " + column + " is [" + str(range_min) + ", " + str(range_max) + "]")
    for index, row in df_merged.iterrows():
        if (row[column] < range_min or row[column] > range_max):
            print("dropping row " + str(index) + " (" + column + " = " + str(row[column]) + ")")
            df_merged = df_merged.drop(index)
df_merged = df_merged.reset_index()

# Print the dataset
df_merged

mean of age is 49.11594202898551
standard deviation of age is 15.837065523714815
acceptable range for age is [1.6047454578410623, 96.62713860012995]
mean of bp is 74.20289855072464
standard deviation of bp is 11.1518890453395
acceptable range for bp is [40.74723141470614, 107.65856568674315]
dropping row 37 (bp = 110.0)
mean of bgr is 130.53284671532847
standard deviation of bgr is 60.59811197791783
acceptable range for bgr is [-51.26148921842503, 312.327182649082]


dropping row 6 (bgr = 424.0)
dropping row 35 (bgr = 380.0)
mean of bu is 52.93333333333333
standard deviation of bu is 47.324534801758944
acceptable range for bu is [-89.0402710719435, 194.90693773861017]
dropping row 15 (bu = 235.0)
dropping row 21 (bu = 309.0)
dropping row 24 (bu = 208.0)
mean of sc is 1.9401515151515154
standard deviation of sc is 2.473654219051764
acceptable range for sc is [-5.480811142003777, 9.361114172306808]
dropping row 9 (sc = 15.0)
dropping row 10 (sc = 11.9)
dropping row 32 (sc = 12.0)
dropping row 33 (sc = 11.8)
mean of sod is 139.4453125
standard deviation of sod is 6.338070608620187
acceptable range for sod is [120.43110067413943, 158.45952432586057]
dropping row 7 (sod = 120.0)
dropping row 19 (sod = 115.0)
dropping row 38 (sod = 114.0)
mean of pot is 4.6848
standard deviation of pot is 3.671157776545417
acceptable range for pot is [-6.32867332963625, 15.698273329636251]
dropping row 20 (pot = 47.0)
mean of hemo is 14.143548387096775
standard deviation

,level_0,index,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0,285,546225,1.0,0.0,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,14600.0,3.2,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
1,1,180,828592,3.0,0.0,71.0,70.0,219.0,82.0,3.6,133.0,4.4,10.4,33.0,5600.0,3.6,0,1,1,0,0,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,1,0
2,3,245,197452,4.0,1.0,56.0,70.0,210.0,26.0,1.7,136.0,3.8,16.1,52.0,12500.0,5.6,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0
3,4,268,923613,4.0,1.0,64.0,60.0,239.0,58.0,4.3,137.0,5.4,9.5,29.0,7500.0,3.4,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,0,1,1,0,1,0
4,5,314,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
113,133,115,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
114,134,112,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
115,135,102,235579,0.0,0.0,46.0,60.0,102.0,27.0,0.7,142.0,4.9,13.2,44.0,11000.0,5.4,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
116,136,149,155117,0.0,0.0,68.0,60.0,125.0,41.0,1.1,139.0,3.8,17.4,50.0,6700.0,6.1,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns.append("al")
numerical_columns.append("su")
# al and su are nominal columns, not numerical, but they range from 1-5 and need to be normalized like everything else.
for column in numerical_columns:
    df_merged[column] = df_merged[column] / df_merged[column].abs().max()

# Print the dataset
df_merged

,level_0,index,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0,285,546225,0.25,0.00,0.5750,0.6,0.537954,0.564417,0.388235,0.940000,0.689655,0.550562,0.518519,0.874251,0.492308,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
1,1,180,828592,0.75,0.00,0.8875,0.7,0.722772,0.503067,0.423529,0.886667,0.758621,0.584270,0.611111,0.335329,0.553846,0,1,1,0,0,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,1,0
2,3,245,197452,1.00,0.25,0.7000,0.7,0.693069,0.159509,0.200000,0.906667,0.655172,0.904494,0.962963,0.748503,0.861538,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0
3,4,268,923613,1.00,0.25,0.8000,0.6,0.788779,0.355828,0.505882,0.913333,0.931034,0.533708,0.537037,0.449102,0.523077,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,0,1,1,0,1,0
4,5,314,343710,0.50,0.00,0.7625,0.8,0.570957,0.907975,0.458824,0.900000,0.896552,0.432584,0.444444,0.550898,0.492308,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
113,133,115,489160,0.00,0.00,0.3500,0.6,0.260726,0.306748,0.058824,0.966667,0.862069,0.988764,0.944444,0.389222,0.769231,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
114,134,112,530138,0.00,0.00,0.2875,0.8,0.366337,0.208589,0.129412,0.966667,0.689655,0.803371,0.759259,0.431138,0.769231,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
115,135,102,235579,0.00,0.00,0.5750,0.6,0.336634,0.165644,0.082353,0.946667,0.844828,0.741573,0.814815,0.658683,0.830769,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
116,136,149,155117,0.00,0.00,0.8500,0.6,0.412541,0.251534,0.129412,0.926667,0.655172,0.977528,0.925926,0.401198,0.938462,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

The "level_0" and "unique_id" columns are useful for data manipulation and identifying individual patients, but they're not useful for modeling and predictions.

In [11]:
#Remove that column
if "level_0" in df_merged:
    df_merged = df_merged.drop(columns = ["level_0", "index", "unique_id"])
# Print the dataset
df_merged

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.25,0.00,0.5750,0.6,0.537954,0.564417,0.388235,0.940000,0.689655,0.550562,0.518519,0.874251,0.492308,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
1,0.75,0.00,0.8875,0.7,0.722772,0.503067,0.423529,0.886667,0.758621,0.584270,0.611111,0.335329,0.553846,0,1,1,0,0,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,1,0
2,1.00,0.25,0.7000,0.7,0.693069,0.159509,0.200000,0.906667,0.655172,0.904494,0.962963,0.748503,0.861538,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0
3,1.00,0.25,0.8000,0.6,0.788779,0.355828,0.505882,0.913333,0.931034,0.533708,0.537037,0.449102,0.523077,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,0,1,1,0,1,0
4,0.50,0.00,0.7625,0.8,0.570957,0.907975,0.458824,0.900000,0.896552,0.432584,0.444444,0.550898,0.492308,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
113,0.00,0.00,0.3500,0.6,0.260726,0.306748,0.058824,0.966667,0.862069,0.988764,0.944444,0.389222,0.769231,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
114,0.00,0.00,0.2875,0.8,0.366337,0.208589,0.129412,0.966667,0.689655,0.803371,0.759259,0.431138,0.769231,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
115,0.00,0.00,0.5750,0.6,0.336634,0.165644,0.082353,0.946667,0.844828,0.741573,0.814815,0.658683,0.830769,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
116,0.00,0.00,0.8500,0.6,0.412541,0.251534,0.129412,0.926667,0.655172,0.977528,0.925926,0.401198,0.938462,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [12]:
# Export the dataframe to a new csv file
df_merged.to_csv('cleaned_data.csv', index=False)
# Export the dataframe to a new json file
df_merged.to_json('cleaned_data.json', index=False)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here
This is what 383GPT outputed:
| al | su | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------|------|-------|------|-------------------------|--------------------------|---------------------------|-------------------------|-------------------------|-------------------------|-------------------------|------------------------|------------------------|---------------|------------|-------------|-----------|----------------|-------------|---------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|-------------|------------------|
| 0.25 | 0.0 | 0.575 | 0.6 | 0.5379537953795379 | 0.5644171779141104 | 0.388235294117647 | 0.94 | 0.6896551724137931 | 0.550561797752809 | 0.5185185185185185 | 0.874251497005988 | 0.49230769230769234 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
| 0.75 | 0.0 | 0.8875| 0.7 | 0.7227722772277227 | 0.5030674846625767 | 0.4235294117647059 | 0.8866666666666667 | 0.7586206896551725 | 0.5842696629213483 | 0.6111111111111112 | 0.33532934131736525 | 0.5538461538461539 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 1.0 | 0.25 | 0.7 | 0.7 | 0.693069306930693 | 0.15950920245398773 | 0.19999999999999998 | 0.9066666666666666 | 0.6551724137931034 | 0.9044943820224719 | 0.9629629629629629 | 0.7485029940119761 | 0.8615384615384615 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
| 1.0 | 0.25 | 0.8 | 0.6 | 0.7887788778877888 | 0.3558282208588957 | 0.5058823529411764 | 0.9133333333333333 | 0.9310344827586208 | 0.5337078651685393 | 0.5370370370370371 | 0.4491017964071856 | 0.5230769230769231 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 |
| 0.5 | 0.0 | 0.7625| 0.8 | 0.570957095709571 | 0.9079754601226994 | 0.4588235294117647 | 0.9 | 0.896551724137931 | 0.43258426966292135 | 0.4444444444444444 | 0.5508982035928144 | 0.49230769230769234 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 |
| 0.5 | 0.0 | 0.7 | 0.9 | 0.42574257425742573 | 0.656441717791411 | 0.788235294117647 | 0.8733333333333333 | 0.8275862068965517 | 0.5112359550561797 | 0.5370370370370371 | 0.38323353293413176 | 0.5230769230769231 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
| 1.0 | 0.0 | 0.075 | 0.6 | 0.3102310231023102 | 0.4110429447852761 | 0.11764705882352941 | 0.9 | 0.8448275862068967 | 0.5561797752808989 | 0.5555555555555556 | 1.0 | 0.7384615384615384 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 |
| 0.75 | 0.0 | 0.775 | 0.7 | 0.40264026402640263 | 0.25766871165644173 | 0.19999999999999998 | 0.9066666666666666 | 0.810344827586207 | 0.7078651685393258 | 0.7222222222222222 | 0.47305389221556887 | 0.6 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
| 1.0 | 0.5 | 0.8 | 1.0 | 0.5379537953795379 | 0.3312883435582822 | 0.8470588235294118 | 0.9333333333333333 | 0.793103448275862 | 0.44382022471910115 | 0.48148148148148145 | 0.4491017964071856 | 0.5230769230769231 |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

It looks like there's too much data for ChatGPT to process, as there are only 9 rows. It might be a good idea to separate this request into subqueries, in this case with 8 or fewer rows per query.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [21]:
# Code to rename all the columns in the dataset
print(categorical_columns)
rename_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_normal': 'Normal Red Blood Cells',
    'rbc_abnormal': 'Abnormal Red Blood Cells',
    'pc_normal': 'Normal Pus Cell',
    'pc_abnormal': 'Abnormal Pus Cell',
    'pcc_present': 'Pus Cell Clumps Present',
    'pcc_notpresent': 'Pus Cell Clumps Not Present',
    'ba_present': 'Bacteria Present',
    'ba_notpresent': 'Bacteria Not Present',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wc': 'White Blood Cell Count',
    'rc': 'Red Blood Cell Count',
    'htn_yes': 'Hypertension',
    'htn_no': 'No Hypertension',
    'dm_yes': 'Diabetes Mellitus',
    'dm_no': 'No Diabetes Mellitus',
    'cad_yes': 'Coronary Artery Disease',
    'cad_no': 'No Coronary Artery Disease',
    'appet_good': 'Good Appetite',
    'appet_poor': 'Poor Appetite',
    'pe_yes': 'Pedal Edema',
    'pe_no': 'No Pedal Edema',
    'ane_yes': 'Anemia',
    'ane_no': 'No Anemia',
    'Target_ckd': 'Target: CKD',
    'Target_nockd': 'Target: No CKD',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count'
}

df_merged.rename(columns=rename_mapping, inplace=True)
df_merged

['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']


,Albumin,Sugar,Age,Blood Pressure,Blood Glucose Random,Blood Urea,Serum Creatinine,Sodium,Potassium,Hemoglobin,Packed Cell Volume,White Blood Cell Count,Red Blood Cell Count,Abnormal Red Blood Cells,Normal Red Blood Cells,Abnormal Pus Cell,Normal Pus Cell,Pus Cell Clumps Not Present,Pus Cell Clumps Present,Bacteria Not Present,Bacteria Present,No Hypertension,Hypertension,No Diabetes Mellitus,Diabetes Mellitus,No Coronary Artery Disease,Coronary Artery Disease,Good Appetite,Poor Appetite,No Pedal Edema,Pedal Edema,No Anemia,Anemia,Target: CKD,Target_notckd
0,0.25,0.00,0.5750,0.6,0.537954,0.564417,0.388235,0.940000,0.689655,0.550562,0.518519,0.874251,0.492308,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
1,0.75,0.00,0.8875,0.7,0.722772,0.503067,0.423529,0.886667,0.758621,0.584270,0.611111,0.335329,0.553846,0,1,1,0,0,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,1,0
2,1.00,0.25,0.7000,0.7,0.693069,0.159509,0.200000,0.906667,0.655172,0.904494,0.962963,0.748503,0.861538,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0
3,1.00,0.25,0.8000,0.6,0.788779,0.355828,0.505882,0.913333,0.931034,0.533708,0.537037,0.449102,0.523077,1,0,1,0,1,0,0,1,0,1,0,1,1,0,0,1,0,1,1,0,1,0
4,0.50,0.00,0.7625,0.8,0.570957,0.907975,0.458824,0.900000,0.896552,0.432584,0.444444,0.550898,0.492308,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
113,0.00,0.00,0.3500,0.6,0.260726,0.306748,0.058824,0.966667,0.862069,0.988764,0.944444,0.389222,0.769231,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
114,0.00,0.00,0.2875,0.8,0.366337,0.208589,0.129412,0.966667,0.689655,0.803371,0.759259,0.431138,0.769231,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
115,0.00,0.00,0.5750,0.6,0.336634,0.165644,0.082353,0.946667,0.844828,0.741573,0.814815,0.658683,0.830769,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
116,0.00,0.00,0.8500,0.6,0.412541,0.251534,0.129412,0.926667,0.655172,0.977528,0.925926,0.401198,0.938462,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [23]:
# Converted SQL to Pandas code
result = df_merged.groupby('Target: CKD').size().reset_index(name='count')
result

,Target: CKD,count
0,0,98
1,1,20


This code returns a dataframe that counts the number of patients with and without CKD.